ocaml/testsuite/tests/typing-modes/portable-contend.ml

```diff
@@ -0,0 +1,249 @@
+                              ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+(* reading immutable field from contended record is fine *)
+let foo (r @ contended) = r.b
+[%%expect{|
+val foo : r @ contended -> string @ contended = <fun>
+|}]
+
+(* Force top level to be uncontended and nonportable *)
+let r @ contended = "hello"
+[%%expect{|
+Line 1, characters 4-27:
+1 | let r @ contended = "hello"
+        ^^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+let x @ portable = "world"
+
+let y @ portable = x
+[%%expect{|
+val x : string = "world"
+Line 3, characters 19-20:
+3 | let y @ portable = x
+                       ^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over writing mutable field gives nonportable *)
+let foo () =
+    let r = {a = "foo"; b = "bar"} in
+    let bar () = r.a <- "hello" in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading mutable field gives nonportable *)
+let foo () =
+    let r = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.a in () in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading immutable field is OK *)
+let foo () =
+    let r @ portable = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.b in () in
+    let _ @ portable = bar in
+    ()
+(* CR zqian: currently mutable(legacy) means all records constructed are nonportable,
+   and the above bar is closing over an nonportable record. Once we allow mutable()
+   syntax, we can test this. *)
+[%%expect{|
+Line 2, characters 23-45:
+2 |     let r @ portable = {a = "foo"; b = "bar"} in
+                           ^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+
+(* TESTING arrays *)
+(* reading/writing to array requires uncontended *)
+let foo (r @ contended) = Array.set r 42 "hello"
+[%%expect{|
+Line 1, characters 36-37:
+1 | let foo (r @ contended) = Array.set r 42 "hello"
+                                        ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+let foo (r @ contended) = Array.get r 42
+[%%expect{|
+Line 1, characters 36-37:
+1 | let foo (r @ contended) = Array.get r 42
+                                        ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+let foo (r @ contended) =
+    match r with
+    | [| x; y |] -> ()
+[%%expect{|
+Line 3, characters 6-16:
+3 |     | [| x; y |] -> ()
+          ^^^^^^^^^^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+
```

Comment:
Yes that's true. Currently we have `mutable(uncontended)` which mean we expect new field value to be `uncontended`.

Reply:
Yes.... but we should still test that we can't put contended values into an array, I think. It's possible that the check for arrays is different than, say, the check for mutable fields.

Reply:
Both, actually. Because the mode of the array must be the same as the mode of its contents, we shouldn't be able to put any contended thing into an array. Unless I've misunderstood.

Reply:
Mutating array is only via stdlib, which we are not gonna modify. Do we still test it?

Reply:
I have checked and they are correct.

Reply:
Are there checked-in tests that reflect this?

Reply:
The topic in question is about modes vs. mutation in general, which is implemented by #2369 . However, in that PR, mutation implies global many shared (for compatibility), which makes the system boring and no extra tests were added.

This PR introduces new axes which makes the `mutable` logic testable. I will add tests.

Reply:
In fact would you mind I create a JIRA ticket to track this instead of doing this in this PR? I don't want unrelated tests in this PR.

Reply:
The test I’m looking for — that you can’t put contended things into an array — seems related to this patch, which introduces the contention axis. But I’m also fine with delaying this piece with an internal ticket.